dds/DCPS/InternalDataReader.h

```diff
@@ -184,62 +192,257 @@ class InternalDataReader : public InternalEntity {
+  {
+    samples.clear();
+    infos.clear();
+
+    ACE_GUARD(ACE_Thread_Mutex, g, mutex_);
+    typename InstanceMap::iterator pos = instance_map_.find(key);
+    if (pos != instance_map_.end()) {
+      pos->second.take(samples, infos);
+      if (pos->second.instance_state() != DDS::ALIVE_INSTANCE_STATE) {
+        instance_map_.erase(pos);
+      }
+    }
+  }
+/// @}
 
 private:
-  const bool durable_;
+  const DDS::DataReaderQos qos_;
   // Often, the listener will have the reader as a member.  Use a weak
   // pointer to prevent a cycle that prevents the listener from being
   // destroyed.
   Listener_wrch listener_;
 
   typedef OPENDDS_SET(InternalEntity_wrch) PublicationSet;
 
-  struct Instance {
-    OPENDDS_LIST(T) samples;
-    OPENDDS_LIST(InternalSampleInfo) infos;
+  class Instance {
+  public:
+
+    Instance()
+      : view_state_(DDS::NEW_VIEW_STATE)
+      , instance_state_(DDS::ALIVE_INSTANCE_STATE)
+      , disposed_generation_count_(0)
+      , no_writers_generation_count_(0)
+    {}
+
+    DDS::ViewStateKind view_state() const { return view_state_; }
+
+    DDS::InstanceStateKind instance_state() const { return instance_state_; }
+
+    void read(SampleSequence& samples, InternalSampleInfoSequence& infos)
+    {
+      size_t sample_count = 0;
+
+      for (typename SampleList::const_iterator pos = not_read_samples_.begin(), limit = not_read_samples_.end();
+           pos != limit; ++pos) {
+        samples.push_back(pos->sample);
+        infos.push_back(make_sample_info(DDS::NOT_READ_SAMPLE_STATE, view_state_, instance_state_, pos->disposed_generation_count, pos->no_writers_generation_count, 0, 0, 0, pos->valid_data));
+        ++sample_count;
+      }
+
+      for (typename SampleList::const_iterator pos = read_samples_.begin(), limit = read_samples_.end();
+           pos != limit; ++pos) {
+        samples.push_back(pos->sample);
+        infos.push_back(make_sample_info(DDS::READ_SAMPLE_STATE, view_state_, instance_state_, pos->disposed_generation_count, pos->no_writers_generation_count, 0, 0, 0, pos->valid_data));
+        ++sample_count;
+      }
+
+      read_samples_.splice(read_samples_.end(), not_read_samples_);
+
+      compute_ranks(sample_count, infos);
+
+      view_state_ = DDS::NOT_NEW_VIEW_STATE;
+    }
+
+    void take(SampleSequence& samples, InternalSampleInfoSequence& infos)
+    {
+      size_t sample_count = 0;
+
+      for (typename SampleList::const_iterator pos = read_samples_.begin(), limit = read_samples_.end();
+           pos != limit; ++pos) {
+        samples.push_back(pos->sample);
+        infos.push_back(make_sample_info(DDS::READ_SAMPLE_STATE, view_state_, instance_state_, pos->disposed_generation_count, pos->no_writers_generation_count, 0, 0, 0, pos->valid_data));
+        ++sample_count;
+      }
+      read_samples_.clear();
+
+      for (typename SampleList::const_iterator pos = not_read_samples_.begin(), limit = not_read_samples_.end();
```

Comment:
Maybe this doesn't matter or I'm not getting something, but `read` goes over `not_read_samples_`, then `read_samples_`. Why is `take` is doing it in the opposite order, `read_samples_`, then `not_read_samples_`?

Reply:
When reading, the goal is to get samples to the read state.
When taking, the goal is to remove samples from the data reader so we start with ones that have already been read.
Ultimately, these methods will take a count and masks that limits the number of samples so the difference will become more relevant.